borsh-derive-internal/src/enum_ser.rs

```diff
@@ -19,6 +23,21 @@ pub fn enum_ser(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
     let mut all_variants_idx_body = TokenStream2::new();
     let mut fields_body = TokenStream2::new();
     let discriminants = discriminant_map(&input.variants);
+    let has_explicit_discriminants = input
+        .variants
+        .iter()
+        .any(|variant| variant.discriminant.is_some());
+    if has_explicit_discriminants && use_discriminant.is_none() {
+        return Err(syn::Error::new(
+            input.ident.span(),
+            "You have to specify `#[borsh(use_discriminant=true)]` or `#[borsh(use_discriminant=false)]` for all structs that have enum with explicit discriminant",
+        ));
+    }
+
+    assert!(
+        input.variants.len() < 256,
+        "up to 256 enum variants are supported"
+    );
     for variant in input.variants.iter() {
```

Comment:
there's no dispatch of behaviour on `use_discriminant` in this file, that's why the minimal tests, mentioned above, fail